furbaby/api/views.py

```diff
@@ -177,3 +187,125 @@ def password_reset_token_created(sender, instance, reset_password_token, *args,
     )
     msg.attach_alternative(email_html_message, "text/html")
     msg.send()
+
+
+@api_view(["GET", "POST", "OPTIONS", "PUT", "PATCH", "DELETE"])
+def user_location_view(request):
+    location_view = UserLocationView()
+
+    if not request.user.is_authenticated:
+        return json_response({"isAuthenticated": False}, status=status.HTTP_401_UNAUTHORIZED)
+
+    # fetch all user locations for the user
+    if request.method == "GET":
+        locations_list = location_view.get_user_locations(request.user.id)
+        return json_response(locations_list, status=status.HTTP_200_OK, safe=False)
+
+    # insert a new location record for the user
+    if request.method in ["POST"]:
+        request.data["user_id"] = request.user.id
+        return location_view.insert_location_record(request.data)
+
+    # update a location record for the user
+    if request.method in ["PUT", "PATCH"]:
+        return location_view.update_location_record(request.user.id, request.data)
+
+    # delete a location record for the user
+    if request.method == "DELETE":
+        return location_view.delete_location_record(request.user.id, request.data)
+
+    return json_response(
+        {"error": "incorrect request method supplied"},
+        status=status.HTTP_405_METHOD_NOT_ALLOWED,
+    )
+
+
+# This class is for the user location(s)
+class UserLocationView(APIView):
+    authentication_classes = []
+
+    # Fetch the locations serializer
+    serializer_class = UserLocationSerializer
+
+    def get_exception_handler(self):
+        return exception_handler
+
+    # takes as input the user id, request and inserts a new location record for the user
+    def insert_location_record(self, request_data):
+        serializer = self.serializer_class(data=request_data)
+
+        if not serializer.is_valid():
+            return json_response(data=serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        instance = serializer.save()
+
+        return json_response(
+            instance.id, status=status.HTTP_201_CREATED, include_data=False, safe=False
+        )
+
+    # takes as input a location object and returns a dictionary of the location key value pairs
+    def get_location_record(self, location=None):
+        if location == None:
+            return None
+        return {
+            "id": location.id,
+            "address": location.address,
+            "city": location.city,
+            "country": location.country,
+            "zipcode": location.zipcode,
+            "user_id": location.user_id,
+            "default_location": location.default_location,
+        }
+
+    # takes as input a user_id and returns a JSON of all the locations for that user
+    def get_user_locations(self, user_id):
+        locations = Locations.objects.filter(user_id=user_id)
+        serialized_data = serialize("json", locations)
+        serialized_data = json.loads(serialized_data)
+        return serialized_data
```

Comment:
let's try something else with this? I see that the response is still a `JSON string` instead of it being a `JSON array`
I'd really appreciate it if you could spend a little bit more time on this and get a better way of providing the response.

Reply:
We could create a list (ie array) of dictionaries containing the locations. Is that an acceptable format for the response?

Reply:
<img width="756" alt="image" src="https://github.com/gcivil-nyu-org/INET-Monday-Fall2023-Team-1/assets/57152208/63daa994-ecb7-4203-9b31-3beab0634e49">

Reply:
yeah, this works for now, I mean...we're not going to be dealing with 1000s of objects where performance of the API is affected at the moment. But in the meanwhile, if we do find something that can work...we should definitely come back to this and replace it! 👍

Reply:
okay